[CU-37t8q47] Added RFC 0013 - Depracate Currency Factory (#2489)
ComposableFi/composable: rfcs/0013-deprecate-currency-factory.md

### Currency Factory Ranges are Confusing
  
The Currency Factory ranges are a confusing abstraction that we don't 
stand to gain from. We really only have two types of currencies native 
(mintable) and non-native (external). However, we currently have many more 
ranges that enforce artificial limitations on various abstractions of these two 
currency types. This complicates the currency creation process and leads to 
confusion while providing no benefit to us.

### No Permissionless Asset Creation

> We do not have permissionless asset (non-payable/no sufficient) - it is 
impossible to create no marketable assets (which is needed for 
governance/airdrops).

### Currency Factory is Already Losing Functionality

Through RFCs 0009 and 0010, we decided to move away from storing ED in Currency 
Factory. After the removal of ED storage, Currency Factory was left with two 
uses: reserving asset IDs and storing asset metadata. 

As discussed, Currency Factory overcomplicates the asset ID reservation system. 
As for asset metadata, we don't have standards for formatting this nor do we 
utilize it. Assuming we replace the reservation system, Currency Factory will be 
left without value.

> ## Solution Proposals

Reviewer: Going directly to solutions without giving any concrete requirements fulfilled by this solution based on the provided reasoning seems odd to me. That means the readers can not actually assess whether a solution is both necessary and sufficient.

# Depreciation of Currency Factory - Assessment RFC

This assessment aims to gather all relevant details and consequences of 
depreciating Currency Factory and its replacements.

## Reasoning

There are several reasons we stand to benefit from replacing Currency Factory.
These reasons are detailed below:

> ### Currency Factory Ranges are Confusing

Reviewer: :+1:

> The Currency Factory ranges are a confusing abstraction that we don't 
> stand to gain from. We really only have two types of currencies native 
> (mintable) and non-native (external). However, we currently have many more 
> ranges that enforce artificial limitations on various abstractions of these two 
> currency types. This complicates the currency creation process and leads to 
> confusion while providing no benefit to us.

Reviewer 2: ```suggestion
The Currency Factory ranges are a confusing abstraction that we don't 
stand to gain from. We really only have two types of currencies native 
(mintable) and non-native (external). However, we currently have many more 
ranges that enforce artificial limitations on various abstractions of these two 
currency types. This complicates the currency creation process and leads to 
confusion while providing no benefit to us. 
Since asset IDs aren't intended for human consumption, there is no requirement for them to be "easily consumable" - the end user should be concerned with the asset's _symbol_, not it's id.
```

Since asset IDs aren't intended for human consumption, there is no requirement for them to be "easily consumable" - the end user should be concerned with the asset's _symbol_, not it's id.

### No Permissionless Asset Creation

> > We do not have permissionless asset (non-payable/no sufficient) - it is 

Reviewer: I'm don't really get this statement, why do we need to create new assets for governance and airdrops?
And what is meant by "no marketable assets"?

Author: If @dzmitry-lahoda could elaborate on this, that would be appreciated

governance/airdrops).

### Currency Factory is Already Losing Functionality

Through RFCs 0009 and 0010, we decided to move away from storing ED in Currency 
Factory. After the removal of ED storage, Currency Factory was left with two 
uses: reserving asset IDs and storing asset metadata. 

As discussed, Currency Factory overcomplicates the asset ID reservation system. 
As for asset metadata, we don't have standards for formatting this nor do we 
utilize it. Assuming we replace the reservation system, Currency Factory will be 

> left without value.

Reviewer 2: are there plans to replace the reservation system? what are possible alternatives?

Reviewer: Good question, this background section is not detailed enough on what other projects are doing for example. Is there a best practice?

Author: I've started a section on what other chains have historically done. Acala section has been pushed with one on Moonbeam being a WIP

left without value.

> ## Requirements

Reviewer: I recommend following [MosCow prioritization](https://www.productplan.com/glossary/moscow-prioritization) to understand if a requirement is actually a requirement while also understanding priority.

If we are to deprecate Currency Factory, we will need to ensure the requirements
it fulfills are still met by some solution.

> * Permissioned asset creation can be done without collision in the asset ID

Reviewer: The requirement is not clear here. Why do we need permissioned asset creation? and through what interface? through an extrinsic or pallet interface trait?

Reviewer: I'm not even sure we need general collision avoidance tbh, We may only need some namespace mechanism such as [MultiLocation](https://polkadot.network/blog/xcm-the-cross-consensus-message-format/#-assets-in-xcm)

> * Asset metadata is available

Reviewer: We need to be more specific here what type of meta data are we talking about?

Additionally, Currency Factory failed to meet some requirements that we should
enforce in a future solution.

* Permissionless asset creation can be done without collision in the asset ID

## Solution Proposals

There are multiple ways we could go about replacing Currency Factory. Possible 

> solutions are not necessarily exclusive (i.e. we could implement multiple). 

Reviewer 2: ```suggestion
solutions are not necessarily mutually exclusive (i.e. we could implement multiple). 
```

> * Permissionless asset creation can be done without collision in the asset ID

Reviewer: The requirement is not clear here. Why do we need permissionless asset creation? and through what interface? through an extrinsic or pallet interface trait?

> 

Reviewer: Are we missing any migration requirement from the existing solution to any new solution?

These solutions are detailed below:

### Simplify & Move the Asset ID Reservation System

One way to remove the need for pallet-currency-factory is to move its 
functionality to pallet-asset-registry. If we also simplify the asset ID 

> reservation system, this will be a minimal change.

Reviewer 2: ```suggestion
reservation system to no longer use predefined ranges, this will be a minimal change.
```

> This keeps the functionality of Currency Factory around but removes the need for
> another pallet. This does not enable premissionless asset creation without the
> need for more design.

Reviewer 2: ```suggestion
This retains the functionality of Currency Factory, and does not require a new pallet. However, this does not enable permissionless asset creation without the
need for more design.
```

Reviewer 2: Not sure what's meant by "more design" here, perhaps it could be rephrased?

Author: We lack a system for creating assets without root/governance
Which may become something we need in the future. Currency Factory is permissioned

Reviewer 2: got it, so if we end up having a requirement for permissionless asset creation we would have to design it

#### Technical Implementation

> Instead of reserving asset IDs via our current range system, we could simply 
> have a nonce for reserving new asset IDs while ensuring a lack of collisions. 
> This nonce can start at an arbitrary but high number so that our hard-coded 
> asset IDs are still safe. This would reduce the complexity of our current 
> reservation system while still avoiding collision.

Reviewer 2: ```suggestion
Instead of reserving asset IDs via our current range system, we could simply 
have one nonce for reserving new asset IDs, which would (by the definition of a nonce) ensure no collisions. 
This nonce can start at an arbitrary but high number so that our hard-coded 
asset IDs are still safe. This would reduce the complexity of our current 
reservation system while still avoiding collision.
```

#### Consequences

* If we already have assets in various ranges, the runtime migration process
will become more complicated as we translate them to existing in one range.
  

> ### Remove Asset ID Reservation

Reviewer 2: What is the purpose of asset id reservation? (this is a gap in my knowledge, but a short explanation/ refresher would be nice to have in this document regardless)

Author: Ensuring new asset IDs don't collide with existing ones

Reviewer 2: that would be solved by using a single nonce right?

Author: It could be, yes

Reviewer 2: This is related to https://github.com/ComposableFi/composable/pull/2489/files#r1037535250, I think that section could be updated with this

Reviewer 2: Maybe not in that section specifically, but somewhere (lol)

> > We do not need [asset ID reservation]. We just runtime config assets to be 

Reviewer 2: what is meant by "We just runtime config assets to be permissioned"?

Author: Meaning that, as long as a permissioned source is creating an asset (root or governance) we are safe to assume that we don't need to automate the ID acquisition process

  * Can run two pallets at same time for two purposes (external and mintable) 
  (Demonstrated by Moonbeam)

  * Handles WASM contracts addresses as IDs

  * Permissionless when needed

  * Supported by Polkadot wallets
  
  * Supports asset metadata
  
  #### Cons
  
  * No asset ID reservation system, only prevents duplicates

  * Hold functionality is not present (We depend on this in several cases)

* **Moonbeam's pallet-asset-manager**

Note: Not looking at using or forking this directly, but instead how they use 
different instances of the same pallet for mintable and external assets.

  #### Pros
  
  * Maintained by moonbeam
  
  #### Cons
  
  * Partially specialized

  * No asset ID reservation system, only prevents duplicates

#### Consequences

* Any of these options will require a difficult runtime migration as we move 
existing storage for pallet-assets/pallet-asset-registry/pallet-currency-factory 
to an alternative.

## Conclusion

After analyzing possible solutions, the best course of action to deprecate 
Currency Factory is to maintain a fork of Parity's pallet-assets, embrace the 
design Moonbeam uses to maintain two instances of their pallet, and to remove 

> our asset ID reservation system (for release two).

Reviewer 2: what benefits does forking bring over simply wrapping the pallet?

Reviewer 2: Could we upstream the changes we need? This may be a more long-term task, so possibly out of scope for what we want to do with this RFC; perhaps a "future possibilities" section would be good?

Author: Up-streaming is an option, probably not one for release two tho.

Reviewer 2: makes sense, i think that should be mentioned as explicitly out of scope then

This can be accomplished in two main phases:

1. Create a fork of Parity's pallet-assets and add the features we need

2. Replace our pallet-assets and pallet-currency-factory with two instances of
Parity's pallet-assets.

Parts of either of these phases can be scaled down for release two.

### Fork of Parity's pallet-assets

1. Create a non-traditional fork of Parity's pallet-assets

2. Implement `frame_support::traits::tokens::fungibles::MutateHold` for 
pallet-assets

3. Translate calls into our old pallet-assets and pallet-currency-factory to 
our new pallet-assets

### Manage Two Instances of pallet-assets

1. Configure instance for mintable assets

2. Configure instance for external assets

> 3. Route calls between the two instances

Author: Looking for comments on this.
I was considering using Assets Registry as a routing layer, but that may be out of scope for Assets Registry.

the asset does not already exist.

Native assets can then be updated by their `CurrencyId`, in which case
the assets `AssetMetadata` can be updated.

#### General Notes

* The `AssetMetadatas` storage map is used for all types and is quarried by the 
`AssetIds` `enum`.

* The `AssetIds` type is also used for routing regular currency functions (i.e. 
transfer, withdraw, deposit).

* For non-erc20 assets, Acala uses 'orml-tokens' for currency functions

#### Data Structures & Types

Note: `MultiLocation` is supplied by XCM.

```rust
pub type ForeignAssetId = u16;
```

```rust
pub struct AssetMetadata<Balance> {
  pub name: Vec<u8>,
	pub symbol: Vec<u8>,
	pub decimals: u8,
	pub minimal_balance: Balance,
}
```

```rust
pub enum AssetIds {
	Erc20(EvmAddress),
	StableAssetId(StableAssetPoolId),
	ForeignAssetId(ForeignAssetId),
	NativeAssetId(CurrencyId),
}
```

```rust
pub enum CurrencyId {
	Token(TokenSymbol),
	DexShare(DexShare, DexShare),
	Erc20(EvmAddress),
	StableAssetPoolToken(StableAssetPoolId),
	LiquidCrowdloan(Lease),
	ForeignAsset(ForeignAssetId),
}
```

### Moonbeam

Moonbeam has two asset classifications: 

* Local
* Foreign

> Moonbeam maintains separate instances of Parity's pallet-assets to track each of

Reviewer: This approach sounds neat for our asset-registry to eventually evolve in to

these asset classifications. Another pallet, pallet-asset-manager, is used to 
route between the two.

Local asset IDs are determined by the result of hashing a counter that tracks
the number of local assets. While Foreign asset IDs are determined by the assets
XCM Multilocation.

#### Foreign

Local assets are registered with a `ForeignAssetType`, `AssetRegisterarMetadata`, a `Balance` representing the ED, and a `bool` `is_sufficent`.

## Requirements

If we are to deprecate Currency Factory, we will need to ensure the requirements
it fulfills are still met by some solution.

* Permissioned asset creation can be done without collision

* Asset metadata is available

> * Can registrar both local and foreign asset types

Reviewer: Can we try to define what assets are considered foreign(assets created by other sovereign chains than the current executing chain) vs local(assets created by the chain that executes the logic). This will be helpful for anyone trying to understand how we handle assets.

Reviewer 3: did not looked, but likely moonbeam uses Call filters to prevent any calls to extrinsic on foreign assets registry to prevent any minting or burning even via sudo.

nonce) ensure no collisions. This nonce can start at an arbitrary but high 
number so that our hard-coded asset IDs are still safe. This would reduce the 
complexity of our current reservation system while still avoiding collision.

#### Consequences

* If we already have assets in various ranges, the runtime migration process
will become more complicated as we translate them to existing in one range.
  
### Remove Asset ID Reservation

> We do not need [asset ID reservation]. We just runtime config assets to be 
permissioned. When we will have permisionless assets they may already have 
reservation (any way making fork with reservation is easy later).

Asset reservation helps us automate new asset IDs. However, for the scope of 
release two - this may not be necessary. We could instead hard-code asset IDs 
for LP tokens and other instances.

Given we can currently manage without automated asset creation, as we can forgo
the asset creation requirements for release two and ensure that all asset data
is still made available.

#### Consequences

* Minor runtime migration will be needed for deleting Currency Factory's 
storage.

> ### Use an Externally Maintained Asset ID Reservation System

Reviewer: This is a separate concern as proposed before, we can focus on concern one for release 2. But not a problem to document the solution. I agree in  concept though we need to assess how well it supports our requirements such as asset locking at the time of implementation(release 3 or after)

  option.
  
  > It is not clear from RFC why substrate assets are good. IMHO it should be 
  clarified - year of man work to make UI/QA/FE/users happy is what we lack. 
  IMHO better to have fork of assets and collaborate with ecosystem to tune 
  wallets. Do not write own wallets.
  
  #### Pros
  
  * Very nice UI compatibility with Polkadot Dashboard

  * Maintained by Parity

  * Simple Interface

  * Can run two pallets at same time for two purposes (external and mintable) 
  (Demonstrated by Moonbeam)

  * Handles WASM contracts addresses as IDs

  * Permissionless when needed

  * Supported by Polkadot wallets
  
  * Supports asset metadata
  
  #### Cons
  
  * No asset ID reservation system, only prevents duplicates

  * Hold functionality is not present (We depend on this in several cases)

* **Moonbeam's pallet-asset-manager**

Note: Not looking at using or forking this directly, but instead how they use 
different instances of the same pallet for mintable and external assets.

  #### Pros
  
  * Maintained by moonbeam
  
  #### Cons
  
  * Partially specialized

  * No asset ID reservation system, only prevents duplicates

#### Consequences

> * Any of these options will require a difficult runtime migration as we move 

Reviewer: I don't think runtime migration is difficult if we design and plan the steps correctly.

existing storage for pallet-assets/pallet-asset-registry/pallet-currency-factory 
to an alternative.

## Conclusion

After analyzing possible solutions, the best course of action to deprecate 

> Currency Factory is to maintain a fork of Parity's pallet-assets, embrace the 

Reviewer: Note that pallet-assets is still a work in progress https://github.com/paritytech/substrate/issues/8453 , we need `fungibles::*Hold` traits for staking in release 3, which pallet-assets does not support. Any solution suggesting adoption of that pallet need to take into account orml-tokens features that we need for release 3. I would rather deal with a migration into pallet-assets(it supports that for genesis https://github.com/paritytech/substrate/blob/master/frame/assets/src/lib.rs#L362) later than delay our feature roadmap. Note that the number of tokens(LPT * 3, PBLO, pool shares * 3) would not be that large even after release 3 and staking. We can even hard code a migration for that. 

I propose the following steps to impl in brief

#### Release 2: 
1. Make sure that we don't use currency-factory and create and track(1) the assets in assets-registry with `MultiLocation` even for local assets.
2. Allow Pablo to create assets(LPT) through assets-registry (avoid use of currency-factory)
3. Make sure one can not use local assets to pay gas fees in XCM transfers (Which seems to be the case AFAIK)
4. For ED for LPT: Asset ratios gets defined in a suitable way at the time of creation for LP (1 LP = value of provided liquidity in PICA / number of tokens)

#### Release 3: 
1. Introduce staking with pool share tokens (Just for context, not related to RFC)
2. Remove currency-factory all together (Just to minimize the changes for release 2)
3. Possibly upgrade substrate to latest production version (assess later).

#### Release 4 or later
1. Assess/Introduce frame/pallet-assets with a migration for existing set of assets if `fungibles::*Hold` trait impls are available.
2. Remove use of our own pallet-assets together with orml-tokens use.

design Moonbeam uses to maintain two instances of their pallet, and to remove 
our asset ID reservation system (for release two).

Additionally, for foreign asset IDs, we can move to the standard of using XCM 
Multi Locations. We can move to using a simple nonce for local asset IDs. Most
of the foreign asset system is already implemented through Assets Registry and
will not need significant changes.

However, we currently lack the means of creating local assets. The proposal here
is to use Assets Registry to register both local and foreign assets. Assets
Registry will act as a routing layer between our two instances of pallet-assets.

This will leave us with two independently stored asset ranges (local and 
foreign) without the overhead of Currency Factory. 

### Technical

1. Asset IDs & Balances

> 

Author: Instead of the complexity of the enum, routing can be handled internally by the pallet by confirm if some storage items exist or not - at the cost of storage access

Neither option presented seems to handle the reserving of asset IDs. Therefore, 
neither of these will fully replace currency factory and would more correctly 
replace our own pallet-assets. A replacement for pallet-assets may be a good 
idea but would require much more substantial changes A replacement for 
pallet-assets may be a good idea but would require much more substantial 
changes.

* **Parity's pallet-assets**
  
  #### Pros
  
  * Very nice UI compatability with Polkadot Dashboard
  
  * Maintained by Parity
  
  * Simple Interface
  
  #### Cons
  
  * No asset ID reservation system, only prevents duplicates
  
  * Hold functionality is not present (We depend on this in several cases)

* **Moonbeam's pallet-asset-manager**

  #### Pros
  
  * Maintained by moonbeam
  
  #### Cons
  
  * Partially specialized
  
  * No asset ID reservation system, only prevents duplicates

## Questions

* > As for asset metadata, we don't have standards for formatting this nor do we 
  utilize it.
  Is this true? A breif investigation revealed that outside of `scripts/polkadot-launch/initialization/src/interfaces/basilisk/definitions.ts` 

>   there is no mention of the Asset Metadata structure.

Author: > As for asset metadata, we don't have standards for formatting this nor do we 
  utilize it.
  
 Is this true? A breif investigation revealed that outside of `scripts/polkadot-launch/initialization/src/interfaces/basilisk/definitions.ts` there is no mention of the Asset Metadata structure.

# Depreciation of Currency Factory - Assessment RFC

This assessment aims to gather all relevant details and consequences of 
depreciating Currency Factory and its replacements.

## Reasoning

There are several reasons we stand to benefit from replacing Currency Factory.
These reasons are detailed below:

### Currency Factory Ranges are Confusing
  

> As detailed in [this](https://composablefinance.slack.com/archives/C031G5NT0CA/p1667492928188269) 

Reviewer 3: please summarize details out of thread here

Reviewer 3: (or you have done it already, just remove it or put into some no so distracting place - so we are not forced to read - i  know I do same way sometimes, but IMHO rfc could be cleaner)

thread, the Currency Factory ranges are a confusing abstraction that we don't 
stand to gain from. We really only have two types of currencies native 

> (mintable) and non-native (external). However, we have many more ranges that 

Reviewer 3: second issue - we do not have permissionless asset (non payable/no sufficitent) - it is impossible to create no marketable assets (which is needed for governance/airdrops)

enforce artificial limitations on various abstractions of these two currency 
types. This complicates the currency creation process and leads to confusion 
while providing no benefit to us.

### Currency Factory is Already Losing Functionality

Recently, we moved away from storing ED in Currency Factory. This leaves two 
uses for currency factory: reserving asset IDs and storing asset metadata. 

> As we have already discussed, currency factory over complicates the asset ID 

Reviewer 3: not sure if this section and paragraph clear (clear for me in general, not sure it is clear for others)

reservation system. As for asset metadata, we don't have standards for 
formatting this nor do we utilize it.

Assuming we replace the reservation system, Currency Factory will be left 
without value.

## Solution Proposals

There are multiple ways we could go about replacing Currency Factory. These 
solutions are detailed below:

### Simplify & Move the Asset ID Reservation System

One way to remove the need for pallet-currency-factory is to move its 

> functionality to pallet-asset-registry. If we also simplify the asset ID 

Reviewer 3: 1. we are 4 month old orml-tokens. what state they have now? may be they did some stuff for non payable assets and substrate assets compatibility?
2. could we move functionally to our assets pallet? can we make our assets pallet to be as compatible as possible with substrate assets?

so in theory we can ditch currency factory and move all to our assets in a way compable on binary level with substrate assets.

also we can throw away our governance-registry imho

Author: > we are 4 month old orml-tokens. what state they have now? may be they did some stuff for non payable assets and substrate assets compatibility?

We'll either need to request another back-port or upgrade our runtime to enjoy their newer additions

> could we move functionally to our assets pallet? can we make our assets pallet to be as compatible as possible with substrate assets?

Like you concluded later, I think maintaining a fork would be the easiest way to get all the benefits of Parity's pallet-assets

Reviewer 2: Is our old substrate version holding us back from using newer features? And if so, what is keeping us on an older substrate version?

reservation system, this will be a minimal change.

Instead of reserving asset IDs via our current range system, we could simply 
have a nonce for reserving new asset IDs while ensuring a lack of collisions. 
This nonce can start at an arbitrary but high number so that our hard-coded 

> asset IDs are still safe (i.e `u32::MAX`). This would reduce the complexity of 

Reviewer 3: Hardcoded could be u16::MAX, u32::MAX could be reserved for permissioned assets. 
So may be we can avoid to be too detailed in this RFC? I mean what is abstractions on top of registry? Just reservation of some subset of ids - we can specify max later.

Imho paragraphs which jump from biz value to tech details are hard to read. May be outline tech proposal later.

Imho using assets is more of biz and ecosystem and general development proposal, then tech.

our current reservation system while still avoiding collision.

### Use an Externally Maintained Asset ID Reservation System

Neither option presented seems to handle the reserving of asset IDs. Therefore, 
neither of these will fully replace currency factory and would more correctly 
replace our own pallet-assets. A replacement for pallet-assets may be a good 
idea but would require much more substantial changes A replacement for 
pallet-assets may be a good idea but would require much more substantial 
changes.

* **Parity's pallet-assets**
  

>   #### Pros

Reviewer 3: - supported by polkadot wallets

  
  * Very nice UI compatibility with Polkadot Dashboard
  
  * Maintained by Parity
  
  * Simple Interface
  
  #### Cons
  

>   * No asset ID reservation system, only prevents duplicates

Reviewer 3: we do not need. we just runtime config assets to be premissioned. when we will have permisionless assets they may already have reservation (any way making fork with reservation is easy later)

Reviewer 3: so really we do not care lack of reservation system

Reviewer 3: now

Author: +1
I've included this in the solution section

>   * Hold functionality is not present (We depend on this in several cases)

Reviewer 3: we can use assets-registry here. it can do next
1. implement orml traits
2. implement substrate traits
3. implement free/reserved/hold/lock in consistent/property tested/fuzzied way
4. route calls to assets

Author: > implement free/reserved/hold/lock in consistent/property tested/fuzzied way

Is there any particular reason that hold and lock/freeze are implemented differently?

> * **Moonbeam's pallet-asset-manager**

Reviewer 3: we should not use assets-manager imho. i used moonbeam just as example of good use of pallet-assets (it is proves assets are good for ETH bridges )

Reviewer 3: pallet-asset-manager - idea we will have our own `assets-registy` which will be tune for our needs.

it is important to check that `pallet-assets` can be wrapped

Reviewer: Agree for not using Moonbeam stuff but we can learn from their impl :)
